tensorflow/compiler/mlir/tensorflow/transforms/interface_optimize_pass.cc

```diff
@@ -0,0 +1,79 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include "mlir/IR/PatternMatch.h"  // from @llvm-project
+#include "mlir/Pass/Pass.h"  // from @llvm-project
+#include "tensorflow/compiler/mlir/tensorflow/ir/tf_traits.h"
+
+namespace mlir {
+namespace TF {
+namespace {
+
+template <typename TraitPattern, template <typename> class Trait>
+struct TraitRewritePattern : public RewritePattern {
+  TraitRewritePattern(MLIRContext* context)
+      : RewritePattern({}, 0, context, MatchAnyOpTypeTag()) {}
+
+  LogicalResult matchAndRewrite(Operation* op,
+                                PatternRewriter& rewriter) const override {
+    if (op->hasTrait<Trait>()) {
+      return static_cast<const TraitPattern*>(this)->matchAndRewriteTrait(
+          op, rewriter);
+    }
+
+    return failure();
+  }
+};
+
+struct FoldInvolution
+    : TraitRewritePattern<FoldInvolution, OpTrait::TF::Involution> {
+  using Base = TraitRewritePattern<FoldInvolution, OpTrait::TF::Involution>;
+
+  FoldInvolution(MLIRContext* context) : Base(context) {}
+
+  LogicalResult matchAndRewriteTrait(Operation* op,
+                                     PatternRewriter& rewriter) const {
+    auto argument_op = op->getOperand(0).getDefiningOp();
+
+    if (argument_op && op->getName() == argument_op->getName()) {
+      // Replace the outer involutions output with inner's input.
+      rewriter.replaceOp(op, argument_op->getOperand(0));
+      return success();
+    }
+
+    return failure();
+  }
```

Comment:
This seems like a useful general canonicalization based on the involution trait, which again seems to be a good fit upstream. Can you submit this there?

Reply:
Do I move the whole PR to MLIR then?

Reply:
Well, almost: we'll need to tag TF ops with the right traits.

Reply:
Where should I put the FoldInvolution upstream? It's tricky because trait matching is based on match any so it doesn't mix well with other patterns. In lib/Transforms/Utils/FoldUtils.cpp I see OperationFolder::tryToFold should I add the body of FoldInvolution there?

Reply:
lib/Transforms/Utils/FoldUtils.cpp seems a bit too "internal" to me right now.

Seems like we would want Traits to be able to extend the folding just like they extend the verifier. Can you look into how "verifierTraits()" is setup and take inspiration from this to propose a solution?

Reply:
So I traced through the trait verifier and its essentially just a separate pass. When you set the verifyPasses flag to true then whenever a pass gets added to the manager another verify pass is created. I also looked at how applyPatternsAndFoldGreedily works which i think is where the "folder" in ops is used and the fold gets applied on each op when it gets iterated over. I think overall it makes most sense to just create a separate pass then for FoldInvolution with PassWrapepr and we can provide it as a lib/Transforms or a directory that makes sense. Separate pass is also general and we can use it for the other trait optimizations. we can provide a pass that combines multiple traits too if that is possible once we implement those.

Reply:
I was referring only to the mechanism by which adding a traits to an op will make the `verifierTraits` function invoked, nothing else: there is a chaining mechanism that make it so that the verifier calls a single method on the operation and that dispatch to all the traits.
We'd need a similar mechanism for dispatch to a fold method implemented by a trait.

The folding hook is a per-op thing that is invoked in various places: a separate pass does not provide what is needed really.

Reply:
oh ok, I think I understand what you mean now. So say I add this mechanism that folds by dispatching to all traits. What would then be the best way to use it? Should I add a pass that for each op calls this dispatcher through the getAbstractOperation, or do we just add the dispatch functionality  and then users will manually call it>

Reply:
You don't need a pass: fold is called by the canonicalize pass.